File: example-clients/python/ws_client.py
#!/usr/bin/env python3
"""Consume a WorldSignal subscription over a WebSocket.

Frames are {"seq": <n>, "payload": <envelope>}. Requires the `websockets`
package (see requirements.txt). Env: same as sse_client.py.
"""
import asyncio
import json
import os
import sys

import websockets

BASE = os.environ.get("WS_API_BASE", "http://localhost:4000")
KEY = os.environ.get("WS_API_KEY", "")
SUB = os.environ.get("WS_SUBSCRIPTION", "demo-stream")
SINCE = os.environ.get("WS_SINCE", "0")
MAX = int(os.environ.get("WS_MAX", "0"))


async def run() -> int:
    if not KEY:
        print("WS_API_KEY is required", file=sys.stderr)
        return 2
    url = BASE.replace("http://", "ws://").replace("https://", "wss://")
    url += f"/v1/stream/ws?subscription={SUB}&since={SINCE}"
    print(f"[ws] connecting to {url}", file=sys.stderr)

    seen = 0
    async with websockets.connect(url, additional_headers={"Authorization": f"Bearer {KEY}"}) as ws:
        async for raw in ws:
            msg = json.loads(raw)
            if msg.get("type") == "connected":
                print(f"[ws] connected to {msg.get('subscription')}", file=sys.stderr)
                continue
            d = msg.get("payload", {}).get("data", {})
            print(f"[ws] {d.get('severity','?'):8} {d.get('country') or '--'}  {d.get('title','')}")
            seen += 1
            if MAX and seen >= MAX:
                return 0
    return 0


if __name__ == "__main__":
    raise SystemExit(asyncio.run(run()))

File: example-clients/node/ws_client.mjs
#!/usr/bin/env node
// Consume a WorldSignal subscription over a WebSocket using Node's built-in
// global WebSocket (Node 21+/22). That client is browser-compatible and can't
// set request headers, so it authenticates with the ?api_key= query param
// (the same fallback the browser uses). Frames are {seq, payload} JSON, with a
// leading {type:"connected"} handshake. Env: same as sse_client.mjs.
const BASE = process.env.WS_API_BASE || "http://localhost:4000";
const KEY = process.env.WS_API_KEY || "";
const SUB = process.env.WS_SUBSCRIPTION || "demo-stream";
const SINCE = process.env.WS_SINCE || "0";
const MAX = Number(process.env.WS_MAX || "0");

if (!KEY) { console.error("WS_API_KEY is required"); process.exit(2); }
const url = BASE.replace(/^http/, "ws") +
  `/v1/stream/ws?subscription=${SUB}&since=${SINCE}&api_key=${encodeURIComponent(KEY)}`;
console.error(`[ws] connecting to ${BASE.replace(/^http/, "ws")}/v1/stream/ws`);

let seen = 0;
const ws = new WebSocket(url);
ws.addEventListener("message", (e) => {
  const msg = JSON.parse(e.data);
  if (msg.type === "connected") { console.error(`[ws] connected to ${msg.subscription}`); return; }
  const d = msg.payload?.data || {};
  console.log(`[ws] ${(d.severity || "?").padEnd(8)} ${d.country || "--"}  ${d.title || ""}`);
  if (MAX && ++seen >= MAX) { ws.close(); process.exit(0); }
});
ws.addEventListener("error", (e) => { console.error("[ws] error", e.message || e); process.exit(1); });
